File: Lab2/Lab2.cpp
#include <iostream>
#include <vector>
#include <fstream>
using namespace std;

class Matrix {
private:
    vector<vector<int>> mat;
    int rows, cols;

public:
    Matrix(int r, int c) : rows(r), cols(c), mat(r, vector<int>(c, 0)) {}

    void input_mat(const string &newfile) {
        ifstream theMat(newfile);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                theMat >> mat[i][j];
            }
        }
        theMat.close(); 
    }
    
    int get_first_element() const {
        return mat[0][0];
    }

    void find_row(int first_num, const string &newfile) {
        int foundRow = -1;
        int maxVal = mat[0][0];
        int minVal = mat[0][0];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (mat[i][j] > maxVal) {
                    maxVal = mat[i][j];
                }
                if (mat[i][j] < minVal) {
                    minVal = mat[i][j];
                }
                if (mat[i][j] == first_num && i != 0 && foundRow == -1) {
                    foundRow = i;
                }
            }
        }

        ofstream outfile(newfile); // ofstream to make the file
        outfile << "-------------------------------------------------------------------------------" << endl;
        if (foundRow != -1) {
            outfile << "Target value in matrix is shown in row " << (foundRow + 1) << endl;
        } else {
            outfile << "Target value in matrix is not shown" << endl;
        }
        outfile << endl << "Maximum value is " << maxVal << endl;
        outfile << "Minimum value is " << minVal << endl;
        outfile << "-------------------------------------------------------------------------------" << endl;
        outfile.close();
    }

    Matrix matrix_addition(const Matrix &next) {
        Matrix result(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result.mat[i][j] = mat[i][j] + next.mat[i][j];
            }
        }
        return result;
    }
    
    Matrix matrix_multiplication(const Matrix &next) {
        Matrix result(rows, next.cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < next.cols; j++) {
                result.mat[i][j] = 0;
                for (int k = 0; k < cols; k++) {
                    result.mat[i][j] += mat[i][k] * next.mat[k][j];
                }
            }
        }
        return result;
    }

    void save(const string &newfile) {
        ofstream outfile(newfile);
        outfile << "-------------------------------------------------------------------------------" << endl;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                outfile << mat[i][j] << " ";
            }
            outfile << endl;
        }
        outfile << endl << "-------------------------------------------------------------------------------" << endl;
        outfile.close();
    }
    
};

int main() {
    int r = 10, c = 10; // Matrix size
    Matrix mat1(r, c), mat2(r, c);
    
    mat1.input_mat("matrix_1.txt");
    mat2.input_mat("matrix_2.txt");
    
    int first_num = mat1.get_first_element();
    mat1.find_row(first_num, "output_Q1_1.txt");
    mat2.find_row(first_num, "output_Q1_2.txt");
    
    Matrix add_result = mat1.matrix_addition(mat2);
    add_result.save("output_Q2.txt");
    
    Matrix mult_result = mat1.matrix_multiplication(mat2);
    mult_result.save("output_Q3.txt");

    
    return 0;
}

File: Lab2/Lab2/Lab2.py
import numpy as np
import pickle

class MatrixOperations:
    def __init__(self, matrix):
        self.matrix = matrix
        self.rows, self.cols = matrix.shape
        
    def find_row(self, first_num):
        max_val = np.max(self.matrix)
        min_val = np.min(self.matrix)
        found_row = -1

        for i in range(1, self.rows):  
            if first_num in self.matrix[i]:
                found_row = i
                break

        return (found_row + 1 if found_row != -1 else -1), max_val, min_val
       
    def matrix_addition(self, other):
        result = np.zeros((self.rows, self.cols), dtype=int)
        for i in range(self.rows):
            for j in range(self.cols):
                result[i][j] = self.matrix[i][j] + other.matrix[i][j]
        return result
        
    def matrix_multiplication(self, other):
        result = np.zeros((self.rows, other.cols), dtype=int)
        for i in range(self.rows):
            for j in range(other.cols):
                for k in range(self.cols):  # Common dimension
                    result[i][j] += self.matrix[i][k] * other.matrix[k][j]
        return result
        
    def save_to_file(self, filename, matrix):
        with open(filename, "w") as f:
            f.write("-------------------------------------------------------------------------------\n")
            for row in matrix:
                f.write(" ".join(map(str, row)) + "\n")
            f.write("-------------------------------------------------------------------------------\n")


if __name__ == "__main__":
    mat1 = np.loadtxt("matrix_1.txt", dtype=int)
    mat2 = np.loadtxt("matrix_2.txt", dtype=int)

    matrix1 = MatrixOperations(mat1)
    matrix2 = MatrixOperations(mat2)

    first_num = mat1[0, 0]
    row_idx_mat1, max_mat1, min_mat1 = matrix1.find_row(first_num)
    row_idx_mat2, max_mat2, min_mat2 = matrix2.find_row(first_num)
    
    with open("output_Q1_1.txt", "w") as f1, open("output_Q1_2.txt", "w") as f2:
        f1.write("-------------------------------------------------------------------------------\n")
        f1.write("Target value in matrix is shown in row {}\n".format(row_idx_mat1 if row_idx_mat1 != -1 else 'not shown'))
        f1.write("Maximum value is {}\n".format(max_mat1))
        f1.write("Minimum value is {}\n".format(min_mat1))
        f1.write("-------------------------------------------------------------------------------\n")


        f2.write("-------------------------------------------------------------------------------\n")
        f2.write("Target value in matrix is shown in row {}\n".format(row_idx_mat2 if row_idx_mat2 != -1 else 'not shown'))
        f2.write("Maximum value is {}\n".format(max_mat2))
        f2.write("Minimum value is {}\n".format(min_mat2))
        f2.write("-------------------------------------------------------------------------------\n")


    mat_add = matrix1.matrix_addition(matrix2)
    matrix1.save_to_file("output_Q2.txt", mat_add)

    mat_mul = matrix1.matrix_multiplication(matrix2)
    matrix1.save_to_file("output_Q3.txt", mat_mul)

    results = {
        'idx_row_mat_1': row_idx_mat1,
        'max_mat_1': max_mat1,
        'min_mat_1': min_mat1,
        'idx_row_mat_2': row_idx_mat2,
        'max_mat_2': max_mat2,
        'min_mat_2': min_mat2,
        'mat_add': mat_add,
        'mat_mul': mat_mul
    }
    
    with open("output_py.pickle", "wb") as f:
        pickle.dump(results, f)

    print("Results saved to output_py.pickle")
